File: backend/app/schemas/rag_debug.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class RetrievalConfigPayload(BaseModel):
    top_k: int = Field(default=8, ge=1, le=50)
    threshold: float = Field(default=0.2, ge=0.0, le=1.0)
    rerank_enabled: bool = True
    chunk_size: int = Field(default=800, ge=100, le=4000)
    chunk_overlap: int = Field(default=100, ge=0, le=2000)
    scope: str = Field(default='knowledge_base', pattern='^(knowledge_base|session)$')


class RetrievalConfigResponse(BaseModel):
    scope: str
    knowledge_base_id: int | None = None
    session_id: int | None = None
    top_k: int
    threshold: float
    rerank_enabled: bool
    chunk_size: int
    chunk_overlap: int
    updated_at: datetime | None = None


class RAGDebugLogItem(BaseModel):
    id: int
    knowledge_base_id: int
    session_id: int | None = None
    user_id: int
    question: str
    rewritten_question: str | None = None
    answer: str | None = None
    retrieval_params: dict[str, Any] | None = None
    retrieved_hits: list[dict[str, Any]] = []
    rerank_before: list[dict[str, Any]] = []
    rerank_after: list[dict[str, Any]] = []
    prompt_context: list[dict[str, Any]] = []
    prompt_text: str | None = None
    citations: list[dict[str, Any]] = []
    generation_stage: str | None = None
    error_message: str | None = None
    total_duration_ms: int | None = None
    rewrite_duration_ms: int | None = None
    embedding_duration_ms: int | None = None
    retrieval_duration_ms: int | None = None
    rerank_duration_ms: int | None = None
    prompt_build_duration_ms: int | None = None
    is_bad_case: bool = False
    bad_case_category: str | None = None
    bad_case_note: str | None = None
    created_at: datetime


class RAGDebugLogListResponse(BaseModel):
    items: list[RAGDebugLogItem]
    total: int
    limit: int
    offset: int


class RAGDebugLogMarkBadCasePayload(BaseModel):
    is_bad_case: bool
    bad_case_category: str | None = Field(default=None, max_length=50)
    bad_case_note: str | None = Field(default=None, max_length=2000)

File: frontend/src/api/ragDebug.ts
import { apiClient } from './client'

export interface RetrievalConfigPayload {
  top_k: number
  threshold: number
  rerank_enabled: boolean
  chunk_size: number
  chunk_overlap: number
  scope: 'knowledge_base' | 'session'
}

export interface RetrievalConfigResponse extends RetrievalConfigPayload {
  knowledge_base_id: number | null
  session_id: number | null
  updated_at: string | null
}

export interface RAGDebugLogItem {
  id: number
  knowledge_base_id: number
  session_id: number | null
  user_id: number
  question: string
  rewritten_question: string | null
  answer: string | null
  retrieval_params: Record<string, any> | null
  retrieved_hits: Array<Record<string, any>>
  rerank_before: Array<Record<string, any>>
  rerank_after: Array<Record<string, any>>
  prompt_context: Array<Record<string, any>>
  prompt_text: string | null
  citations: Array<Record<string, any>>
  generation_stage: string | null
  error_message: string | null
  total_duration_ms: number | null
  rewrite_duration_ms: number | null
  embedding_duration_ms: number | null
  retrieval_duration_ms: number | null
  rerank_duration_ms: number | null
  prompt_build_duration_ms: number | null
  is_bad_case: boolean
  bad_case_category: string | null
  bad_case_note: string | null
  created_at: string
}

export interface RAGDebugLogListResponse {
  items: RAGDebugLogItem[]
  total: number
  limit: number
  offset: number
}

export const getRetrievalConfig = async (knowledgeBaseId: number, sessionId?: number | null) => {
  const { data } = await apiClient.get<RetrievalConfigResponse>('/rag-debug/config', {
    params: {
      knowledge_base_id: knowledgeBaseId,
      session_id: sessionId || undefined,
    },
  })
  return data
}

export const saveRetrievalConfig = async (knowledgeBaseId: number, payload: RetrievalConfigPayload, sessionId?: number | null) => {
  const { data } = await apiClient.put<RetrievalConfigResponse>('/rag-debug/config', payload, {
    params: {
      knowledge_base_id: knowledgeBaseId,
      session_id: sessionId || undefined,
    },
  })
  return data
}

export const listRagDebugLogs = async (params?: {
  knowledge_base_id?: number
  session_id?: number
  question?: string
  is_bad_case?: boolean
  bad_case_category?: string
  stage?: string
  start_at?: string
  end_at?: string
  limit?: number
  offset?: number
}) => {
  const { data } = await apiClient.get<RAGDebugLogListResponse>('/rag-debug/logs', { params })
  return data
}

export const markRagDebugBadCase = async (
  logId: number,
  payload: { is_bad_case: boolean; bad_case_category?: string | null; bad_case_note?: string | null },
) => {
  const { data } = await apiClient.patch<RAGDebugLogItem>(`/rag-debug/logs/${logId}/bad-case`, payload)
  return data
}

export const exportRagBadCases = async (knowledgeBaseId?: number) => {
  const { data } = await apiClient.get('/rag-debug/logs/export', {
    params: {
      knowledge_base_id: knowledgeBaseId || undefined,
    },
    responseType: 'blob',
  })
  return data as Blob
}
